NetKet/Machine/fullconn_layer.hpp

```diff
@@ -0,0 +1,257 @@
+
+    std::string buffer = "";
+
+    InfoMessage(buffer) << "Fully Connected Layer " << in_size_ << " --> "
+                        << out_size_ << std::endl;
+    InfoMessage(buffer) << "# # UseBias = " << usebias_ << std::endl;
+  }
+
+  void to_json(json &pars) const override {
+    json layerpar;
+    layerpar["Name"] = "FullyConnected";
+    layerpar["UseBias"] = usebias_;
+    layerpar["Inputs"] = in_size_;
+    layerpar["Outputs"] = out_size_;
+    layerpar["Bias"] = bias_;
+    layerpar["Weight"] = weight_;
+
+    pars["Machine"]["Layers"].push_back(layerpar);
+  }
+
+  void from_json(const json &pars) override {
+    if (FieldExists(pars, "Weight")) {
+      weight_ = pars["Weight"];
+    } else {
+      weight_.setZero();
+    }
+    if (FieldExists(pars, "Bias")) {
+      bias_ = pars["Bias"];
+    } else {
+      bias_.setZero();
+    }
+  }
+
+  void InitRandomPars(int seed, double sigma) override {
+    VectorType par(npar_);
+
+    netket::RandomGaussian(par, seed, sigma);
+
+    SetParameters(par, 0);
+  }
+
+  int Npar() const override { return npar_; }
+
+  int Ninput() const override { return in_size_; }
+
+  int Noutput() const override { return out_size_; }
+
+  void GetParameters(VectorType &pars, int start_idx) override {
+    int k = start_idx;
+
+    if (usebias_) {
+      std::memcpy(pars.data() + k, bias_.data(), out_size_ * scalar_bytesize_);
+      k += out_size_;
+    }
+
+    std::memcpy(pars.data() + k, weight_.data(),
+                in_size_ * out_size_ * scalar_bytesize_);
+  }
+
+  void SetParameters(const VectorType &pars, int start_idx) override {
+    int k = start_idx;
+
+    if (usebias_) {
+      std::memcpy(bias_.data(), pars.data() + k, out_size_ * scalar_bytesize_);
+
+      k += out_size_;
+    }
+
+    std::memcpy(weight_.data(), pars.data() + k,
+                in_size_ * out_size_ * scalar_bytesize_);
+  }
+
+  void InitLookup(const Eigen::VectorXd &v, LookupType &lt) override {
+    if (lt.VectorSize() == 0) {
+      lt.AddVector(out_size_);
+    }
+    lt.V(0) = (weight_.transpose() * v + bias_);
+  }
+
+  void UpdateLookup(const Eigen::VectorXd &v, const std::vector<int> &tochange,
+                    const std::vector<double> &newconf,
+                    LookupType &lt) override {
+    if (tochange.size() != 0) {
+      for (std::size_t s = 0; s < tochange.size(); s++) {
+        const int sf = tochange[s];
+        lt.V(0) += weight_.row(sf) * (newconf[s] - v(sf));
+      }
+    }
+  }
+
+  void Forward(const VectorType &prev_layer_data) override {
+    // Linear term z = W' * in + b
+    z_ = bias_;
+    z_.noalias() += weight_.transpose() * prev_layer_data;
+
+    // Apply activation function
+    activation_(z_, a_);
+  }
+
+  // Using lookup
+  void Forward(const VectorType & /*prev_layer_data*/,
+               const LookupType &lt) override {
+    z_ = lt.V(0);
+    // Apply activation function
+    activation_(z_, a_);
+  }
+
+  VectorType Output() const override { return a_; }
```

Comment:
Similarly to my comment on `BackPropData`, it seems to me that this method is redundant, since a_ should just be returned by the two Forward methods. Also, this is potentially error prone, because here one is implicitly assuming that every time one calls `Output` a call to `Forward` has been performed before. What if this is not the case?

Reply:
then i would need to store the output of a each layer in the machine class: FFNN.
You prefer that?

Reply:
I think that would make more sense, yes (and actually even be slightly faster). We should just be careful when returning those vectors though, I am not sure Eigen supports move semantics. It might be a better option to pass the output by reference to the `Forward` method, though less elegant. Maybe @femtobit or @twesterhout know better.

Reply:
Actually I am only now realising that returning a_ by value might be **very** bad for performance in the current implementation.

Reply:
But I dont really like that the layers dont know its own output

Reply:
The layer shouldn't necessarily know about the output, since the output depends on the specific input, that somebody else is feeding to the layer. It's the FFNN that knows both input and output... it seems consistent to me. It's like implementing a class that computes sin(x). You have two options, say: 

```c++
class Sin{

public :

double operator() (double x){
   return sin(x); 
}
}; 
````

versus 
```c++
class Sin{
double y_; 
public :

void operator() (double x){
   y_=sin(x); 
}

double val(){
  return y_; 
}
}; 
```
I think the first option makes much more sense, and it's also safer! (imagine that you call the method Value() before calling the operator () in the second case...

Reply:
ok i will change it to pass by reference then

Reply:
should i also store the value of the output before applying the activation in FFNN?

So that FFNN stores the output, value before activation and din_?